cache-filter/src/unique_callout.rs

```diff
@@ -0,0 +1,129 @@
+    );
+
+    // check if lock is already acquired or not
+    match get_shared_data(&request_key)? {
+        (_, None) => {
+            // lock not acquired yet, try to set lock.
+            // Note: Since two threads can get None for CAS at the same time, to avoid
+            // overwriting of lock by another thread, use any value for CAS other than 0.
+            match set_shared_data(&request_key, Some(root_id_str_bytes), Some(1)) {
+                Ok(()) => {
+                    info!(
+                        context_id,
+                        "thread(id: {}): callout-lock acquired for request(key: {})",
+                        root_id,
+                        request_key
+                    );
+                    Ok(true)
+                }
+                Err(Status::CasMismatch) => {
+                    info!(
+                        context_id,
+                        "thread(id: {}): callout-lock for request(key:{}) already acquired by another thread",
+                        root_id,
+                        request_key
+                    );
+                    Ok(false)
+                }
+                Err(e) => Err(e),
+            }
+        }
+        (Some(bytes), Some(_)) => {
+            let lock_acquired_by =
+                std::str::from_utf8(&bytes).unwrap_or("thread id failed to deserialize");
+            info!(
+                context_id,
+                "thread(id: {}): callout-lock for request(key: {}) already acquired by thread(id: {})",
+                root_id,
+                request_key,
+                lock_acquired_by
+            );
+            Ok(false)
+        }
+        (None, Some(cas)) => {
+            info!(
+                context_id,
+                "thread(id: {}): callout-lock was already free and trying to acquire now for request(key: {})",
+                root_id,
+                request_key
+            );
+            match set_shared_data(&request_key, Some(root_id_str_bytes), Some(cas)) {
+                Ok(()) => {
+                    info!(
+                        context_id,
+                        "thread(id: {}): callout-lock acquired for request(key: {})",
+                        root_id,
+                        request_key
+                    );
+                    Ok(true)
+                }
+                Err(Status::CasMismatch) => {
+                    info!(
+                        context_id,
+                        "thread(id: {}): callout-lock for request(key: {}) already acquired by another thread",
+                        root_id,
+                        request_key
+                    );
+                    Ok(false)
+                }
+                Err(e) => Err(e),
+            }
+        }
+    }
+}
+
+// NOTE: Right now, there is no option of deleting the pair instead only the value can be erased,
+// and it requires changes in the ABI so change this because it will lead to better memory usage.
```

Comment:
One important issue here is whether the Proxy-WASM specification really goes into detail around this. Unfortunately it is very far from giving these details, so we need to work with the implementation we have in Envoy - always note these details to clearly separate the concrete Envoy implementation which is part of our assumptions from the spec, which might not define at all how these details interact with our client code.

Reply:
I think adding a big comment with granular details helps here as well but if you want to re-iterate some SDK info here then I can do that.